learning/courses/quantum-machine-learning/data-encoding.ipynb

```diff
@@ -0,0 +1,1713 @@
+    "\n",
+    "feature_dim = 2\n",
+    "fig, axs = plt.subplots(9, 2)\n",
+    "i_plot = 0\n",
+    "for paulis in [\n",
+    "    [\"I\"],\n",
+    "    [\"X\"],\n",
+    "    [\"Y\"],\n",
+    "    [\"Z\"],\n",
+    "    [\"XX\"],\n",
+    "    [\"XY\"],\n",
+    "    [\"XZ\"],\n",
+    "    [\"YY\"],\n",
+    "    [\"YZ\"],\n",
+    "    [\"ZZ\"],\n",
+    "    [\"X\", \"ZZ\"],\n",
+    "    [\"Y\", \"ZZ\"],\n",
+    "    [\"Z\", \"ZZ\"],\n",
+    "    [\"X\", \"YZ\"],\n",
+    "    [\"Y\", \"YZ\"],\n",
+    "    [\"Z\", \"YZ\"],\n",
+    "    [\"YY\", \"ZZ\"],\n",
+    "    [\"XY\", \"ZZ\"],\n",
+    "]:\n",
+    "    pfm = PauliFeatureMap(feature_dimension=feature_dim, paulis=paulis, reps=1)\n",
+    "    circuit_drawer(\n",
+    "        pfm.decompose(),\n",
+    "        output=\"mpl\",\n",
+    "        style={\"backgroundcolor\": \"#EEEEEE\"},\n",
+    "        ax=axs[int((i_plot - i_plot % 2) / 2), i_plot % 2],\n",
+    "    )\n",
+    "    axs[int((i_plot - i_plot % 2) / 2), i_plot % 2].title.set_text(paulis)\n",
+    "    i_plot += 1\n",
+    "\n",
+    "fig.set_figheight(16)\n",
+    "fig.set_figwidth(16)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "61e585cc-8e84-496e-9e51-21a818731be8",
+   "metadata": {},
+   "source": [
+    "The above can, of course, be extended to include other permutations and repetitions of Pauli matrices. Learners are encouraged to experiment with those options."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "44a27c71-acb8-4b4f-a344-836cae1e3896",
+   "metadata": {},
+   "source": [
+    "## Review of built-in feature maps\n",
+    "\n",
+    "You have seen several schemes for encoding data into a quantum circuit:\n",
+    "- Basis encoding\n",
+    "- Amplitude encoding\n",
+    "- Angle encoding\n",
+    "- Phase encoding\n",
+    "- Dense encoding\n",
+    "\n",
+    "You have seen how to construct your own feature maps using these encoding schemes, and you have seen four built-in feature maps which take advantage of angle and/or phase encoding:\n",
+    "- EfficientSU2\n",
+    "- ZFeatureMap\n",
+    "- ZZFeatureMap\n",
+    "- PauliFeatureMap\n",
+    "\n",
+    "These built-in feature maps differed from each other in several ways:\n",
+    "- The depth for a given number of encoded features\n",
+    "- The number of qubits required for a given number of features\n",
+    "- The degree of entanglement (obviously related to the other differences)\n",
+    "\n",
+    "The code below applies these four built-in feature maps to the encoding of a feature set, and plots the 2-qubit depth of the resulting circuit. Since 2-qubit error rates are much higher than single-qubit gate error rates, one might reasonably be most interested in the depth of 2-qubit gates. In the code below, we obtain counts of all gates in a circuit by first decomposing the circuit and then using count_ops(), as shown below. Here the 2-qubit gates we are interested in are 'cx' gates:"
```

Comment:
```suggestion
    "The code below applies these four built-in feature maps to the encoding of a feature set, and plots the two-qubit depth of the resulting circuit. Since two-qubit error rates are much higher than single-qubit gate error rates, one might reasonably be most interested in the depth of two-qubit gates. In the code below, we obtain counts of all gates in a circuit by first decomposing the circuit and then using count_ops(), as shown below. Here the two-qubit gates we are interested in are 'cx' gates:"
```